File: clusters/casd_spark.R

.CASD <- new.env()
.CASD$jobDt < - NULL

# configuration : lue dans la base de registre
casd_conf < - function()
{
if (.Platform$OS.type != "windows")
stop("Ce helper est prevu pour Windows.")
r < -
try(utils::
    readRegistry("Software\\CASD\\Hadoop", hive="HCU"), silent = TRUE)
if (inherits(r, "try-error") | | is.null(r$ToolsPath)) {
stop("Configuration absente dans HKCU\\Software\\CASD\\Hadoop.\n",
"Executer install-tokens.ps1 avant d'utiliser ce helper.")
}
r
}

# token jetable propre a la session R
casd_token_session < - function()
{
cf < - casd_conf()
dt < - file.path(Sys.getenv("TEMP"), sprintf("hadoop-r-%d.dt", Sys.getpid()))
ps < - file.path(cf$ToolsPath, "refresh-tokens.ps1")
if (!file.exists(ps)) stop("refresh-tokens.ps1 introuvable dans ", cf$ToolsPath)

sortie < - system2("powershell",
                   c("-ExecutionPolicy", "Bypass", "-File", shQuote(ps),
                     "-Out", shQuote(dt), "-Quiet"),
                   stdout=TRUE, stderr=TRUE)
if (!file.exists(dt)) {
stop("Echec de la generation du token :\n", paste(sortie, collapse = "\n"))
}
Sys.setenv(HADOOP_TOKEN_FILE_LOCATION=dt)
.CASD$jobDt < - dt
message("Token Spark genere : ", dt)
invisible(dt)
}

# connexion

casd_spark_connect < - function(config=NULL,
master = "yarn",
app_name = "rstudio",
driver_port = NULL) {

if (!requireNamespace("sparklyr", quietly = TRUE)) {
stop("Le paquet sparklyr n'est pas installe.")
}
cf < - casd_conf()

if (nzchar(cf$SparkHome))  Sys.setenv(SPARK_HOME      = cf$SparkHome)
if (nzchar(cf$HadoopConf)) Sys.setenv(HADOOP_CONF_DIR = cf$HadoopConf)

if (master == "yarn") casd_token_session()

cfg < - if (is.null(config)) sparklyr::spark_config() else config
if (is.null(driver_port)) {
driver_port < - if (! is.null(cf$DriverPort)) as.integer(cf$DriverPort) else 7077L
}

#
# Configuration imposee : securite, placement, ports.
cfg$spark.security.credentials.hadoopfs.enabled < - "false"
cfg$spark.security.credentials.hive.enabled < - "false"
cfg$spark.security.credentials.hbase.enabled < - "false"
cfg$spark.yarn.stagingDir < - paste0(cf$StagingDir, "/", Sys.getenv("USERNAME"))
cfg$spark.driver.port < - driver_port
cfg$spark.driver.blockManager.port < - driver_port + 200

sparklyr::spark_connect(master=master,
                        app_name=app_name,
                        config=cfg,
                        spark_home=Sys.getenv("SPARK_HOME"))
}

# deconnexion et menage
casd_spark_disconnect < - function(sc)
{
try(sparklyr::
    spark_disconnect(sc), silent = TRUE)
if (! is.null(.CASD$jobDt)) {
crc < - file.path(dirname(.CASD$jobDt), paste0(".", basename(.CASD$jobDt), ".crc"))
unlink(c(.CASD$jobDt, crc))
.CASD$jobDt < - NULL
}
# on rend la main au token de session, utilise par les commandes hdfs
cf < -
try(casd_conf(), silent = TRUE)
if (!inherits(cf, "try-error")) {
sess < - file.path(cf$TokenDir, "")  # le token de session est nomme par PID PowerShell
Sys.unsetenv("HADOOP_TOKEN_FILE_LOCATION")
}
invisible(TRUE)
}

# menage automatique si la session R se ferme sans disconnect explicite
if (exists(".CASD", inherits=FALSE)) {
reg.finalizer(.CASD, function(e)
{
if (! is.null(e$jobDt)) unlink(e$jobDt)
}, onexit = TRUE)
}

File: clusters/casd_spark.py
import os
import subprocess
import tempfile
import atexit
import winreg
from contextlib import contextmanager

_REG = r"Software\CASD\Hadoop"
_job_dt = None


def conf_registre():
    """Lit la configuration deposee par install-tokens.ps1."""
    try:
        k = winreg.OpenKey(winreg.HKEY_CURRENT_USER, _REG)
    except OSError:
        raise RuntimeError(
            "Configuration absente dans HKCU\\%s.\n"
            "Executer install-tokens.ps1 avant d'utiliser ce module." % _REG
        )
    valeurs = {}
    i = 0
    while True:
        try:
            nom, val, _ = winreg.EnumValue(k, i)
            valeurs[nom] = val
            i += 1
        except OSError:
            break
    winreg.CloseKey(k)
    return valeurs


def _token_session():
    """Genere un token jetable propre a cette session Python."""
    global _job_dt
    cf = conf_registre()
    dt = os.path.join(tempfile.gettempdir(), f"hadoop-py-{os.getpid()}.dt")
    ps = os.path.join(cf["ToolsPath"], "refresh-tokens.ps1")
    if not os.path.exists(ps):
        raise FileNotFoundError(f"refresh-tokens.ps1 introuvable dans {cf['ToolsPath']}")

    res = subprocess.run(
        ["powershell", "-ExecutionPolicy", "Bypass", "-File", ps,
         "-Out", dt, "-Quiet"],
        capture_output=True, text=True
    )
    if not os.path.exists(dt):
        raise RuntimeError(
            "Echec de la generation du token.\n"
            f"stdout: {res.stdout}\nstderr: {res.stderr}"
        )

    os.environ["HADOOP_TOKEN_FILE_LOCATION"] = dt
    _job_dt = dt
    print(f"Token Spark genere : {dt}")
    return dt


def _cleanup():
    """Supprime le token jetable de la session Python."""
    global _job_dt
    if _job_dt:
        crc = os.path.join(os.path.dirname(_job_dt),
                           "." + os.path.basename(_job_dt) + ".crc")
        for f in (_job_dt, crc):
            try:
                os.remove(f)
            except OSError:
                pass
        _job_dt = None
    os.environ.pop("HADOOP_TOKEN_FILE_LOCATION", None)


atexit.register(_cleanup)


def get_spark(conf=None, master="yarn", app_name="python", driver_port=None):
    """
    Ouvre une SparkSession configuree pour le cluster.

    conf : dictionnaire de proprietes Spark choisies par l'utilisateur
           (memoire, nombre d'executeurs, options applicatives).
           Les cles de securite et de placement ci-dessous sont imposees
           et ecrasent toute valeur fournie.
    """
    cf = conf_registre()

    if cf.get("SparkHome"):
        os.environ["SPARK_HOME"] = cf["SparkHome"]
    if cf.get("HadoopConf"):
        os.environ["HADOOP_CONF_DIR"] = cf["HadoopConf"]

    if master == "yarn":
        _token_session()

    from pyspark.sql import SparkSession
    b = SparkSession.builder.master(master).appName(app_name)

    # configuration de l'utilisateur, appliquee en premier
    for cle, val in (conf or {}).items():
        b = b.config(cle, str(val))

    if driver_port is None:
        driver_port = int(cf.get("DriverPort", 7077))

    b = (b.config("spark.security.credentials.hadoopfs.enabled", "false")
          .config("spark.security.credentials.hive.enabled", "false")
          .config("spark.security.credentials.hbase.enabled", "false")
          .config("spark.yarn.stagingDir",
                  f"{cf['StagingDir']}/{os.environ.get('USERNAME','')}")
          .config("spark.driver.port", str(driver_port))
          .config("spark.driver.blockManager.port", str(driver_port + 200)))
    # -----------------------------------------------------------------

    return b.getOrCreate()


def stop_spark(spark):
    """Ferme la session et fait le menage."""
    try:
        spark.stop()
    finally:
        _cleanup()


@contextmanager
def spark_session(**kwargs):
    """Gestionnaire de contexte : ferme et nettoie meme en cas d'erreur."""
    spark = get_spark(**kwargs)
    try:
        yield spark
    finally:
        stop_spark(spark)